### PR TITLE
Add tests for the meta-cl framework and minor bug fixes

### PR DESCRIPTION
**Pull request by mvacaporale**
This unit tests for taking gradients of gradients. Specifically, it ensure the gradients that flow through the updates in `MetaContinualLearningExperiment.update_params` work as expected.

It also adds regression tests for task samplers used in the meta-cl experiment class. As regression tests, these will need to be manually ran every once in a while.

Note, our fsx file storage that hosts imagenet now hosts Omniglot to support these tests. I've also included a copy MNIST. This will be helpful if we want to make more extensive end-to-end tests of the vernon framework with an easy-to-use dataset.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/datasets/torchvision_factory.py [hunk 1]
```diff
@@ -64,7 +65,7 @@ def torchvisiondataset(root, dataset_name, train=True, download=True,
                 )
                 raise
 
-    with FileLock(f"{root}.lock"):
+    with FileLock(f"{root}.lock") if download else nullcontext():
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/datasets/torchvision_factory.py:68]: Cool, just learned a new trick!

File: nupic/research/frameworks/pytorch/models/common_models.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def forward(self, x):
 
 class OmniglotCNN(nn.Module):
 
-    def __init__(self, num_classes, **kwargs):
+    def __init__(self, num_classes):
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/models/common_models.py:52]: Now I see why you removed `input_size` from the experiments file. Accepting **kwargs make it easier to switch back and forth between networks without having to change model_args as well. 

I am not opposed to it, but it is an experiments framework, and I think as such we should do everything that makes it easier for users to experiment it with. Having consistent APIs for common classes, like networks, is one of these things that makes it easier to perform experiments (for example I can grid search between 3 different networks).

File: nupic/research/frameworks/pytorch/models/common_models.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def conv_block(cls, in_channels, out_channels, kernel_size):
 
 class OMLNetwork(nn.Module):
 
-    def __init__(self, num_classes, **kwargs):
+    def __init__(self, num_classes):
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/models/common_models.py:87]: Same comment as above.

File: tests/unit/frameworks/vernon/meta_continual_learning_test.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+from nupic.research.frameworks.continual_learning.maml_utils import clone_model
+from nupic.research.frameworks.vernon import MetaContinualLearningExperiment
+
+# Retrieve function that updates params in place.
+# This enables taking gradients of gradients.
+update_params = MetaContinualLearningExperiment.update_params
+
+
+class Quadractic(torch.nn.Module):
```
Comment (lucasosouza) [inline tests/unit/frameworks/vernon/meta_continual_learning_test.py:35]: typo: Quadratic

File: tests/unit/frameworks/vernon/meta_continual_learning_test.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        out = torch.matmul(self.weight, x)
+        out = torch.matmul(self.weight.transpose(1, 0), out)
+        return out
+
+
+class GradsOfGradsTest(unittest.TestCase):
+    """
+    Perform tests for taking gradients of gradients. Specifically, this uses a
+    quadratic layer as a test example since it yields non-trivial results (unlike a
+    linear layer) and enables hand derived expected values.
```
Comment (lucasosouza) [inline tests/unit/frameworks/vernon/meta_continual_learning_test.py:55]: Maybe "manually calculated expected values"?

File: tests/unit/frameworks/vernon/meta_continual_learning_test.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        # Input passed to quadratic function: a W^T W x
+        self.left_input = torch.tensor([[1., 1.]], requires_grad=True)  # a
+        self.right_input = torch.tensor([[0.32], [0.72]], requires_grad=True)  # x
+
+        # Expected gradient of gradient for quadratic layer updated via SGD.
+        self.expected_grad = torch.tensor([[0.3972, 0.6762], [0.2869, 0.4458]])
+
+    def test_auto_grad_with_quadratic_function(self):
+        """
+        Test use of pytorch's autograd to keep track of gradients of gradients.
```
Comment (lucasosouza) [inline tests/unit/frameworks/vernon/meta_continual_learning_test.py:68]: There are some formulas written as comments here in this function. If there is an external reference, maybe add the link here or further explain it. 

It is just a test, so completely optional, but it would help someone maintaining the code in the future.

**Review by lucasosouza — CHANGES_REQUESTED** [review]
Request changes only due to the typo in Quadractic. Left some other comments as well.

File: nupic/research/frameworks/pytorch/models/common_models.py (hunk 2 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/common_models.py:52]: It seems odd to include unused arguments. I agree that experimenting should be made easy, but the arguments should reflect the scope and intention of the class.

I'd argue that maybe this network shouldn't be in frameworks. If there are still iterations to make on the network structure, then maybe this should be in projects.

File: tests/unit/frameworks/vernon/meta_continual_learning_test.py (hunk 6 above)
Comment (mvacaporale) [reply tests/unit/frameworks/vernon/meta_continual_learning_test.py:68]: Good point. I included the derivations in the comments.

File: nupic/research/frameworks/pytorch/models/common_models.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/common_models.py:52]: What I believe to be a better approach is instead of throwing yet another error, let it run and just warn the user: "Arguments x, y and z are not being used in this network". The user is probably aware of it already.

I would agree there is probably better solutions, although I can't come up with one now. I just don't think making it even harder to run experiments is a good idea.
